docs: server routes streaming, configuration access, query access, etc (#4401)
nuxt/framework: docs/content/2.guide/2.features/9.server-routes.md

```ts [/server/api/hello.ts]
export default defineEventHandler((event) => {
  return {

>     api: "works",

Reviewer: ```suggestion
    api: 'works',
```

>   };

Reviewer: ```suggestion
  }
```

> });

Reviewer: ```suggestion
})
```

**Example:**

```ts [/server/routes/hello.ts]

> export default defineEventHandler(() => "Hello World!");

Reviewer: ```suggestion
export default defineEventHandler(() => 'Hello World!')
```

```ts [/server/middleware/log.ts]
export default defineEventHandler((event) => {

>   console.log("New request: " + event.req.url);
> });

Reviewer: ```suggestion
  console.log('New request: ' + event.req.url)
})
```

```

```ts [/server/middleware/auth.ts]
export default defineEventHandler((event) => {

>   event.context.auth = { user: 123 };
> });

Reviewer: ```suggestion
  event.context.auth = { user: 123 }
})
```

Handle file names can be suffixed with `.get`, `.post`, `.put`, `.delete`, ... to match request's [HTTP Method](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods).

```ts [/server/api/test.get.ts]

> export default defineEventHandler(() => "Test get handler");

Reviewer: ```suggestion
export default defineEventHandler(() => 'Test get handler')
```

```

```ts [/server/api/test.post.ts]

> export default defineEventHandler(() => "Test post handler");

Reviewer: ```suggestion
export default defineEventHandler(() => 'Test post handler')
```

**Examples:**

```ts [/server/api/foo/[...].ts]

> export default defineEventHandler(() => `Default foo handler`);

Reviewer: ```suggestion
export default defineEventHandler(() => `Default foo handler`)
```

```

```ts [/server/api/[...].ts]

> export default defineEventHandler(() => `Default api handler`);

Reviewer: ```suggestion
export default defineEventHandler(() => `Default api handler`)
```

```

## Handling Requests with Body

```ts [/server/api/submit.post.ts]
export default defineEventHandler(async (event) => {

>   const body = await useBody(event);
>   return { body };
> });

Reviewer: ```suggestion
  const body = await useBody(event)
  return { body }
})
```

We are using `submit.post.ts` in the filename only to match requests with `POST` method that can accept the request body. When using `useBody` within a GET request, `useBody` will throw a `405 Method Not Allowed` HTTP error.
::

## Handling Requests with Query parameters

Sample query `/api/query?param1=a&param2=b`

```ts [/server/api/query.get.ts]
export default defineEventHandler((event) => {
  const query = useQuery(event)
  return { a: query.param1, b: query.param2 }
})
```

## Get access to configuration from nuxt.config.ts

```ts [/server/api/foo.ts]

export default defineEventHandler((event) => {
  const config = useRuntimeConfig()
  return { key: config.KEY }
})
```

## Handling Requests with async response 

```ts [/server/api/foo.get.ts]

> import { send, sendError } from 'h3'

Reviewer 2: Are the import needed here @pi0 ?

Reviewer 3: No. Also no need to use either. Returning response implies sending it and `throw` sends an error + stops the handler process.

Reviewer 3: I think just removing it. We don't need either send or sendError.

We are using `submit.post.ts` in the filename only to match requests with `POST` method that can accept the request body. When using `useBody` within a GET request, `useBody` will throw a `405 Method Not Allowed` HTTP error.
::

## Handling Requests with Query parameters

Sample query `/api/query?param1=a&param2=b`

```ts [/server/api/query.get.ts]
export default defineEventHandler((event) => {
  const query = useQuery(event)
  return { a: query.param1, b: query.param2 }
})
```

## Handling Requests with Streaming

```ts [/server/api/foo.get.ts]
import { sendStream } from 'h3'
import fs from 'fs'

export default defineEventHandler((event) => {

>   return sendStream(event, fs.createReadStream('/path/to/file'))

Reviewer 3: We shall move this section to the last, mention it is only available for Node.js and with experimental support. Internal calls will be stock currently with `sendStrsm`!

})
```

## Handling Requests with proxy

```ts [/server/api/foo.get.ts]
import { send, sendError } from 'h3'
import fs from 'fs'

export default defineEventHandler((event) => {

>   return $fetch('https://example.com', { method: 'GET', parseResponse: JSON.parse })

Reviewer 3: Nitro will directly support Proxy. This is probably not a good practice to mention in examples...

>   return $fetch('https://example.com', { method: 'GET', parseResponse: JSON.parse })
>         .then((r) => {
>             const response = { data : r.data }
>             //[....]
>             send(event, response)
>         })
>         .catch(ex => sendError(event, ex))
>   )

Reviewer: ```suggestion
  return $fetch('https://example.com', { method: 'GET', parseResponse: JSON.parse })
    .then((res) => {
      const response = { data : res.data }
      // ...
      send(event, response)
    })
    .catch(err => sendError(event, err))
  )
```

})
```

## Get access to configuration from nuxt.config.ts

```ts [/server/api/foo.ts]

> import { useRuntimeConfig } from '#nitro'

Reviewer 3: This can be removed we have auto import support

> const config = useRuntimeConfig()

Reviewer 3: Config access should be _within handler_ not global context. This is why it is a function to be always fresh.

We are using `submit.post.ts` in the filename only to match requests with `POST` method that can accept the request body. When using `useBody` within a GET request, `useBody` will throw a `405 Method Not Allowed` HTTP error.
::

## Handling Requests with Query parameters

Sample query `/api/query?param1=a&param2=b`

```ts [/server/api/query.get.ts]

> export default defineEventHandler(async (event) => {
>   const query = useQuery(event);
>   return { a: query.param1, b: query.param2 };
> });

Reviewer: ```suggestion
export default defineEventHandler(async (event) => {
  const query = useQuery(event)
  return { a: query.param1, b: query.param2 }
})
```

```

## Handling Requests with Streaming

```ts [/server/api/foo.get.ts]

> import { sendStream } from "h3";
> import fs from "fs";
> 
> export default defineEventHandler(async (event) => {
>   return sendStream(event, fs.createReadStream("/path/to/file"));
> });

Reviewer: ```suggestion
import { sendStream } from 'h3'
import fs from 'fs'

export default defineEventHandler(async (event) => {
  return sendStream(event, fs.createReadStream('/path/to/file'))
})
```

```

## Handling Requests with proxy

```ts [/server/api/foo.get.ts]

> import { send, sendError } from "h3";
> import fs from "fs";
> 
> export default defineEventHandler(async (event) => {
>     return $fetch("https://example.com", { method: "GET", parseResponse: JSON.parse })
>         .then((r) => {
>             const response = { data : r.data }
>             //[....]
>             send(event, response)
>         })
>         .catch(ex => sendError(event, ex))
>   );

Reviewer: ```suggestion
import { send, sendError } from 'h3'
import fs from 'fs'

export default defineEventHandler(async (event) => {
    return $fetch('https://example.com', { method: 'GET', parseResponse: JSON.parse })
        .then((r) => {
            const response = { data : r.data }
            // [....]
            send(event, response)
        })
        .catch(ex => sendError(event, ex))
  )
```

})
```

## Get access to configuration from nuxt.config.ts

```ts [/server/api/foo.ts]

> import { useRuntimeConfig } from "#nitro";
> 
> const config = useRuntimeConfig();
> 
> export default defineEventHandler(async (event) => {
>   return { key: config.KEY };
> });

Reviewer: ```suggestion
import { useRuntimeConfig } from '#nitro'

const config = useRuntimeConfig()

export default defineEventHandler(async (event) => {
  return { key: config.KEY }
})
```

```

## Access Request Cookies

```ts
export default defineEventHandler((event) => {

>   const cookies = useCookies(event);
>   return { cookies };
> });

Reviewer: ```suggestion
  const cookies = useCookies(event)
  return { cookies }
})
```

```

## Using a nested router

```ts [/server/api/hello.ts]

> import { createRouter } from "h3";

Reviewer: ```suggestion
import { createRouter } from 'h3'
```

> const router = createRouter();

Reviewer: ```suggestion
const router = createRouter()
```

> router.get("/", () => "Hello World");

Reviewer: ```suggestion
router.get('/', () => 'Hello World')
```

> export default router;

Reviewer: ```suggestion
export default router
```

```

## Return a legacy handler or middleware

```ts [/server/api/legacy.ts]
export default (req, res) => {

>   res.end("Legacy handler");
> };

Reviewer: ```suggestion
  res.end('Legacy handler')
}
```

```ts [/server/middleware/legacy.ts]
export default (req, res, next) => {

>   console.log("Legacy middleware");
>   next();
> };

Reviewer: ```suggestion
  console.log('Legacy middleware')
  next()
}
```
